### PR TITLE
Fix adding batch dim

### PR DESCRIPTION
**Pull request by Atticus1806**
cf. #1057 

There probably is a different location to fix this, but I think its a good starting point for a discussion on how to fix this.

### CODE REVIEW

**Comment by albertz** [comment]
Can you also add the test case?

**Comment by albertz** [comment]
This PR is here not really about the flattening logic but about fixing this batch heuristic, right?

**Comment by albertz** [comment]
The change from this PR already fixes the problem?

File: returnn/tf/util/data.py [hunk 1]
```diff
@@ -3471,6 +3471,7 @@ def copy_add_dim_by_tag(self, dim_tag, unbroadcast=False, axis=None):
     if dim_tag.is_batch_dim():
       if unbroadcast:
         batch_info = dim_tag.src_data.batch if dim_tag.src_data else None
+        batch_info = batch_info or dim_tag.batch
         return self.copy_add_batch_dim(batch_dim_axis=axis, batch=batch_info, dim_tag=dim_tag)
```
Comment (albertz) [inline returnn/tf/util/data.py:3475]: This looks fine in general. I would even go further and say that `dim_tag.batch` should be preferred when it is set. Maybe we even don't need the other code at all anymore.
```suggestion
        return self.copy_add_batch_dim(batch_dim_axis=axis, batch=dim_tag.batch, dim_tag=dim_tag)
```

**Comment by albertz** [comment]
> This PR fixes the problem that I for now only found in the flatten logic. I am a bit hesitant since I am changing a function "outside" the flatten logic

But the issue is not at all about the flatten logic?

**Comment by Atticus1806** [comment]
I think it _can_ happen outside the flatten logic, but the way the `dim_tag` is created in the flatten logic this definitely happens.

File: tests/test_TFNetworkLayer.py [hunk 2]
```diff
@@ -7281,6 +7281,48 @@ def test_reduce_with_flatten():
     net.get_total_loss()
 
 
+def test_double_flatten_loss():
+  from returnn.tf.util.data import batch_dim, SpatialDim, FeatureDim
+  time_dim = SpatialDim("time")
+  _repeat_out_dim = time_dim * 5
+  feature_dim = FeatureDim("feat", 1)
+  config = Config({"extern_data": {"data": {"dim_tags": (batch_dim, time_dim, feature_dim), "dtype": "int32"}},
+    "behavior_version": 12}, )
```
[the remark this answers is not in the document]
Comment (Atticus1806) [reply tests/test_TFNetworkLayer.py:7290]: Okay. This is also done in the `test_reduce_with_flatten` Test, even though this PR is not related, is it fine to change here too? Or make a separate PR?

[the remark this answers is not in the document]
Comment (albertz) [reply tests/test_TFNetworkLayer.py]: Separate, or even push this single change directly to master.

**Review by albertz — APPROVED** [review]

**Pull request merged**